File: 2975. maximum-square-area-by-removing-fences-from-a-field.py
from typing import List


class Solution:
    def maximizeSquareArea(self, m: int, n: int, hFences: List[int], vFences: List[int]) -> int:
        hFences.extend([1, m])
        vFences.extend([1, n])

        stt = set()
        ans = 0

        for i in range(len(hFences)):
            for j in range(i+1, len(hFences)):
                stt.add(abs(hFences[j]-hFences[i]))
        
        for i in range(len(vFences)):
            for j in range(i+1, len(vFences)):
                val = abs(vFences[j]-vFences[i])
                if val in stt:
                    ans = max(ans, val)
        
        if ans == 0:
            return -1
        return (ans*ans)%(10**9+7)


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Test case 1
    m1, n1 = 4, 3
    hFences1 = [2, 3]
    vFences1 = [2]
    output1 = solution.maximizeSquareArea(m1, n1, hFences1, vFences1)
    print(f"Test 1: m={m1}, n={n1}, hFences={[2,3]}, vFences={[2]}")
    print(f"Output: {output1}, Expected: 4")
    
    # Test case 2
    m2, n2 = 6, 7
    hFences2 = [2]
    vFences2 = [4]
    output2 = solution.maximizeSquareArea(m2, n2, hFences2, vFences2)
    print(f"Test 2: m={m2}, n={n2}, hFences={[2]}, vFences={[4]}")
    print(f"Output: {output2}, Expected: -1")
    
    print("\nTest cases executed!")

File: 2975. maximum-square-area-by-removing-fences-from-a-field.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int maximizeSquareArea(int m, int n, int[] hFences, int[] vFences) {
        // Extend arrays with boundaries
        int[] hExtended = new int[hFences.length + 2];
        int[] vExtended = new int[vFences.length + 2];
        
        hExtended[0] = 1;
        hExtended[1] = m;
        for (int i = 0; i < hFences.length; i++) {
            hExtended[i + 2] = hFences[i];
        }
        
        vExtended[0] = 1;
        vExtended[1] = n;
        for (int i = 0; i < vFences.length; i++) {
            vExtended[i + 2] = vFences[i];
        }
        
        Set<Integer> stt = new HashSet<>();
        int ans = 0;
        
        // Calculate all possible horizontal distances
        for (int i = 0; i < hExtended.length; i++) {
            for (int j = i + 1; j < hExtended.length; j++) {
                stt.add(Math.abs(hExtended[j] - hExtended[i]));
            }
        }
        
        // Find maximum common distance with vertical fences
        for (int i = 0; i < vExtended.length; i++) {
            for (int j = i + 1; j < vExtended.length; j++) {
                int val = Math.abs(vExtended[j] - vExtended[i]);
                if (stt.contains(val)) {
                    ans = Math.max(ans, val);
                }
            }
        }
        
        if (ans == 0) {
            return -1;
        }
        
        long result = (long) ans * ans;
        return (int) (result % (1_000_000_007));
    }

    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        int m1 = 4, n1 = 3;
        int[] hFences1 = {2, 3};
        int[] vFences1 = {2};
        int output1 = solution.maximizeSquareArea(m1, n1, hFences1, vFences1);
        System.out.println("Test 1: m=4, n=3, hFences=[2,3], vFences=[2]");
        System.out.println("Output: " + output1 + ", Expected: 4");
        
        // Test case 2
        int m2 = 6, n2 = 7;
        int[] hFences2 = {2};
        int[] vFences2 = {4};
        int output2 = solution.maximizeSquareArea(m2, n2, hFences2, vFences2);
        System.out.println("Test 2: m=6, n=7, hFences=[2], vFences=[4]");
        System.out.println("Output: " + output2 + ", Expected: -1");
        
        System.out.println("\nTest cases executed!");
    }
}
